prover/src/logup_gkr/mod.rs

```diff
@@ -0,0 +1,441 @@
+}
+
+impl<E: FieldElement> EvaluatedCircuit<E> {
+    /// Creates a new [`EvaluatedCircuit`] by evaluating the circuit where the input layer is
+    /// defined from the main trace columns.
+    pub fn new(
+        main_trace_columns: &impl Trace<BaseField = E::BaseField>,
+        evaluator: &impl LogUpGkrEvaluator<BaseField = E::BaseField>,
+        log_up_randomness: &[E],
+    ) -> Result<Self, GkrProverError> {
+        let mut layer_polys = Vec::new();
+
+        let mut current_layer =
+            Self::generate_input_layer(main_trace_columns, evaluator, log_up_randomness);
+        while current_layer.num_wires() > 1 {
+            let next_layer = Self::compute_next_layer(&current_layer);
+
+            layer_polys.push(CircuitLayerPolys::from_circuit_layer(current_layer));
+
+            current_layer = next_layer;
+        }
+
+        Ok(Self { layer_polys })
+    }
+
+    /// Returns all layers of the evaluated circuit, starting from the input layer.
+    ///
+    /// Note that the return type is a slice of [`CircuitLayerPolys`] as opposed to
+    /// [`CircuitLayer`], since the evaluated layers are stored in a representation which can be
+    /// proved using GKR.
+    pub fn layers(self) -> Vec<CircuitLayerPolys<E>> {
+        self.layer_polys
+    }
+
+    /// Returns the numerator/denominator polynomials representing the output layer of the circuit.
+    pub fn output_layer(&self) -> &CircuitLayerPolys<E> {
+        self.layer_polys.last().expect("circuit has at least one layer")
+    }
+
+    /// Evaluates the output layer at `query`, where the numerators of the output layer are treated
+    /// as evaluations of a multilinear polynomial, and similarly for the denominators.
+    pub fn evaluate_output_layer(&self, query: E) -> (E, E) {
+        let CircuitLayerPolys { numerators, denominators } = self.output_layer();
+
+        (numerators.evaluate(&[query]), denominators.evaluate(&[query]))
+    }
+
+    // HELPERS
+    // -------------------------------------------------------------------------------------------
+
+    /// Generates the input layer of the circuit from the main trace columns and some randomness
+    /// provided by the verifier.
+    fn generate_input_layer(
+        main_trace: &impl Trace<BaseField = E::BaseField>,
+        evaluator: &impl LogUpGkrEvaluator<BaseField = E::BaseField>,
+        log_up_randomness: &[E],
+    ) -> CircuitLayer<E> {
+        let num_fractions = evaluator.get_num_fractions();
+        let mut input_layer_wires =
+            Vec::with_capacity(main_trace.main_segment().num_rows() * num_fractions);
+        let mut main_frame = EvaluationFrame::new(main_trace.main_segment().num_cols());
+
+        let mut query = vec![E::BaseField::ZERO; evaluator.get_oracles().len()];
+        let mut numerators = vec![E::ZERO; num_fractions];
+        let mut denominators = vec![E::ZERO; num_fractions];
+        for i in 0..main_trace.main_segment().num_rows() {
```

Comment:
My thinking was that, at the "application" level, the selectors will be designed so as to be zero for such edge cases. Thus I didn't give much thought to a choice we can make with respect to `NextRow` oracles, namely either we choose a circular shift, as we do now, or we use a plain shift where we will get zero at the last row as a result.

Reply:
> namely either we choose a circular shift, as we do now, or we use a plain shift where we will get zero at the last row as a result.

Right, we currently do a circular shift. But we actually want a plain shift, and I don't see how this would be possible with the current API. For example, when `x` represents the last row, how does `evaluate_query()` know that ultimately it should output 0s for the numerators and 1s for the denominator (since all it gets is the evaluations of the oracles at `x`)?

I'm currently not seeing it, but if you're confident that it's possible, then we can move on.

Reply:
For example, the numerators could be an expression involving selectors which are zero once a `HALT` instruction has been initiated. For this row, and all subsequent rows assuming that a `HALT` can only be followed by a `HALT`, the numerators will be zero. The denominators are not important as far as I can see, of course as long as they are not zero, and this is guaranteed with high probability by the LogUp randomness.
I might be missing something extra but we can, as you suggested, fix it in a subsequent targeted PR.